src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -344,133 +344,312 @@ public class SqlMutationEngine : IMutationEngine
+                bool isDatabasePolicyDefinedForReadAction = false;
 
-                try
+                if (isReadPermissionConfiguredForRole)
                 {
-                    // Creating an implicit transaction
-                    using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType())
-                    {
-                        upsertOperationResult = await PerformUpsertOperation(
-                                                            parameters,
-                                                            context);
-                        transactionScope.Complete();
-                    }
+                    isDatabasePolicyDefinedForReadAction = !string.IsNullOrWhiteSpace(_authorizationResolver.GetDBPolicyForRequest(context.EntityName, roleName, EntityActionOperation.Read));
                 }
 
-                // All the exceptions that can be thrown by .Complete() and .Dispose() methods of transactionScope
-                // derive from TransactionException. Hence, TransactionException acts as a catch-all.
-                // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
-                // a DataApiBuilderException is thrown
-                catch (TransactionException)
+                if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
                 {
-                    throw _dabExceptionWithTransactionErrorMessage;
-                }
+                    DbResultSet? upsertOperationResult;
+                    DbResultSetRow upsertOperationResultSetRow;
+                    JsonDocument? selectOperationResponse = null;
+                    bool isUpdateResultSet = false;
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                    try
+                    {
+                        // Creating an implicit transaction
+                        using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType())
+                        {
+                            upsertOperationResult = await PerformUpsertOperation(
+                                                                parameters,
+                                                                context);
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
-                {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
+                            if (upsertOperationResult is null)
+                            {
+                                // Ideally this case should not happen, however may occur due to unexpected reasons,
```

Comment:
Previously, this case with upsertOperationResult=null was being returned back as is with no exceptions.. I understand this shouldn't happen - but will throwing an exception now, ensure it will be appropriately caught and handled?

Reply:
Yes. In the same method, the code block for insert operation throws similar exceptions for null result rows and for 0 rows returned from the database.